src/spaces/openstreetmap.jl

```diff
@@ -0,0 +1,591 @@
+integration example: [Social networks with LightGraphs.jl](@ref).
+
+## The OSMAgent
+
+The base properties for an agent residing on an `OSMSpace` are as follows:
+```julia
+mutable struct OSMAgent <: AbstractAgent
+    id::Int
+    pos::Tuple{Int,Int,Float64}
+    route::Vector{Int}
+    destination::Tuple{Int,Int,Float64}
+end
+```
+
+Current `pos`ition and `destination` tuples are represented as
+`(start intersection index, finish intersection index, distance travelled in meters)`.
+The `route` is an ordered list of intersections, providing a path to reach `destination`.
+
+Further details can be found in [`OSMAgent`](@ref).
+
+## Routing
+
+There are three ways to generate a route, depending on the situation.
+1. Using [`add_agent!`](@ref)`(start, finish, model)`, which will calculate a route automatically for a new agent.
+2. [`osm_plan_route`](@ref), which provides `:shortest` and `:fastest` paths (with the option of a `return_trip`) between intersections or positions.
+3. [`osm_random_route!`](@ref), choses a new `destination` an plans a new path to it; overriding the current route (if any).
+"""
+function OpenStreetMapSpace(
+    path::AbstractString;
+    use_cache = false,
+    trim_to_connected_graph = true,
+)
+    m = get_map_data(path; use_cache, trim_to_connected_graph)
+    agent_positions = [Int[] for i in 1:nv(m.g)]
+    return OpenStreetMapSpace(m, findall(!iszero, m.w), agent_positions)
+end
+
+function Base.show(io::IO, s::OpenStreetMapSpace)
+    print(
+        io,
+        "OpenStreetMapSpace with $(length(s.m.roadways)) roadways and $(length(s.m.intersections)) intersections",
+    )
+end
+
+const OSMSpace = OpenStreetMapSpace
+
+const TEST_MAP =
+    joinpath(dirname(pathof(OpenStreetMapX)), "..", "test", "data", "reno_east3.osm")
+
+#######################################################################################
+# Custom functions for OSMSpace
+#######################################################################################
+
+# EXPORTED
+"""
+    osm_random_road_position(model::ABM{OpenStreetMapSpace})
+
+Similar to `random_position`, but rather than providing only intersections, this method
+returns a location somewhere on a road heading in a random direction.
+"""
+function osm_random_road_position(model::ABM{<:OpenStreetMapSpace})
+    edge = rand(model.space.edges)
+    return (edge.I..., rand() * model.space.m.w[edge])
+end
+
+"""
+    osm_random_route!(agent, model::ABM{<:OpenStreetMapSpace})
+
+Selects a random destination and plans a route from the agent's current position. Will
+overwrite any current route.
+"""
+function osm_random_route!(agent, model::ABM{<:OpenStreetMapSpace})
+    agent.destination = osm_random_road_position(model)
+    agent.route = osm_plan_route(agent.pos, agent.destination, model)
+    return nothing
+end
+
+"""
+    osm_plan_route(start, finish, model::ABM{<:OpenStreetMapSpace}; by = :shortest, return_trip = false, kwargs...)
```

Comment:
docstring is too large to be rendered in the docs

Reply:
Also, what are `kwargs...`?

Reply:
> Road speeds are needed for this method which can be passed in via extra keyword arguments. Consult the OpenStreetMapX documentation for more details.